frontend/src/app/services/resource.service.ts

```diff
@@ -0,0 +1,64 @@
+import { Injectable } from '@angular/core';
+import {
+  HttpClient,
+  HttpErrorResponse,
+  HttpParams,
+} from '@angular/common/http';
+import { environment } from '../../environments/environment';
+import { Observable, throwError } from 'rxjs';
+import { catchError, map, switchMap, take } from 'rxjs/operators';
+import { IResponse } from '../core/models/response';
+
+@Injectable({
+  providedIn: 'root',
+})
+export abstract class ResourceService<T> {
+  private readonly APIUrl = environment.apiUrl + this.getResourceUrl();
+
+  constructor(protected httpClient: HttpClient) {}
+
+  abstract getResourceUrl(): string;
+
+  getListPagination(index: number, page: number): Observable<T[]> {
+    return this.httpClient.get<IResponse>(`/${this.APIUrl}`).pipe(
+      map((item) => JSON.parse(item.data)),
+      catchError(this.handleError)
+    );
+  }
+
+  getList(): Observable<T[]> {
+    return this.httpClient.get<IResponse>(`/${this.APIUrl}`).pipe(
+      map((item) => JSON.parse(item.data)),
+      catchError(this.handleError)
+    );
+  }
+
+  get(id: string | number): Observable<T> {
+    return this.httpClient.get<IResponse>(`/${this.APIUrl}/${id}`).pipe(
+      map((item) => JSON.parse(item.data)),
+      catchError(this.handleError)
+    );
+  }
+
+  add(resource: T): Observable<IResponse> {
+    return this.httpClient
+      .post<IResponse>(`/${this.APIUrl}`, JSON.stringify(resource))
```

Comment:
You're posting T (the resource). No need to use the IResponse interface here. It's only for responses coming from the backend

Reply:
I edited the code, but the studio is cursing, any ideas
![3](https://user-images.githubusercontent.com/55750629/153704769-7a2327c3-26b3-4e45-8bfc-ae68cd30aab2.png)
?

Reply:
Sure, you should specify that you're posting a resource of a specific type: `.post<T>`. In that way you can even remove this part: `Observable<T>`. TypeScript is smart enough to know what to expect.